File: 백준/Silver/16139. 인간－컴퓨터 상호작용/인간－컴퓨터 상호작용.py
import sys
s=sys.stdin.readline().rstrip()
q=int(sys.stdin.readline().rstrip())
arr=[[0]*len(s) for _ in range(26)]
for j in range(len(s)):
    idx=ord(s[j])-97
    for i in range(26):
        arr[i][j]=arr[i][j-1]
        if i==idx:
            arr[i][j]+=1
for _ in range(q):
    ch,start,end=sys.stdin.readline().rstrip().split()
    idx=int(ord(ch)-97)
    start,end=int(start),int(end)
    r=arr[idx][end]
    if start:
        r-=arr[idx][start-1]
    print(r)

File: 백준/Silver/16139. 인간－컴퓨터 상호작용/인간－컴퓨터 상호작용.cc
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int main(void) {
    int q, len, idx, start = 0, end = 0;
    char *s = (char*)malloc(sizeof(char) * 200001), ch;
    scanf("%s", s);
    scanf("%d", &q);
    len = strlen(s);
    int** arr = (int**)malloc(sizeof(int*) * 26);
    for (int i = 0; i < 26; i++) {
        arr[i] = (int*)malloc(sizeof(int) * len);
        for (int j = 0; j < len; j++) {
            arr[i][j] = 0;
        }
    }
    for (int j = 0; j < len; j++) {
        idx = s[j] - 97;
        for (int i = 0; i < 26; i++) {
            if (j > 0)
                arr[i][j] = arr[i][j - 1];
            if (i == idx)
                arr[i][j] += 1;
        }
    }
    for (int i = 0; i < q; i++) {
        scanf(" %c %d %d", &ch, &start, &end);
        idx = ch - 97;
        len = arr[idx][end];
        if (start)
            len -= arr[idx][start - 1];
        printf("%d\n", len);
    }
    return 0;
}